File: game_cs/src/Pipe.cs
namespace Flappy;

/// <summary>A vertical pipe pair scrolling left, with a gap to fly through.</summary>
public sealed class Pipe : GameObject
{
    public float X;
    public readonly float GapY;
    public readonly float Width = Config.PipeWidth;
    public readonly float Gap = Config.PipeGap;
    public bool Scored;

    public Pipe(float x, float gapY)
    {
        X = x;
        GapY = gapY;
    }

    /// <summary>The two solid rectangles (top and bottom) used for collision.</summary>
    public (Rect top, Rect bottom) Solids(float screenH)
    {
        float topH = GapY - Gap / 2f;
        float bottomY = GapY + Gap / 2f;
        float bottomH = screenH - Config.GroundHeight - bottomY;
        return (
            new Rect(X, 0f, Width, topH),
            new Rect(X, bottomY, Width, bottomH)
        );
    }

    public bool OffScreen => X + Width < 0f;

    public override bool IsAlive => !OffScreen;

    public override void Update(float dt) => X -= Config.ScrollSpeed * dt;

    public override void Draw()
    {
        var (top, bottom) = Solids(Engine.Height);
        foreach (var solid in new[] { top, bottom })
        {
            Engine.DrawRectangle(solid.X, solid.Y, solid.W, solid.H, Color.Green);
            Engine.DrawRectangleLines(solid.X, solid.Y, solid.W, solid.H, 4f, Color.DarkGreen);
        }

        // Classic wider "lip" at each side of the gap.
        const float lipH = 26f;
        const float overhang = 6f;
        float topLipY = GapY - Gap / 2f - lipH;
        float bottomLipY = GapY + Gap / 2f;
        foreach (float lipY in new[] { topLipY, bottomLipY })
        {
            Engine.DrawRectangle(X - overhang, lipY, Width + overhang * 2f, lipH, Color.Green);
            Engine.DrawRectangleLines(X - overhang, lipY, Width + overhang * 2f, lipH, 4f, Color.DarkGreen);
        }
    }
}

File: game_rs/src/pipe.rs
//! Vertical pipe pairs — mirror of `Flappy.Pipe` in C#.

use macroquad::prelude::*;

use crate::config;
use crate::primitives::{PackedColor, Rect};

/// A vertical pipe pair scrolling left, with a gap to fly through.
pub struct Pipe {
    pub x: f32,
    pub gap_y: f32,
    pub width: f32,
    pub gap: f32,
    pub scored: bool,
}

impl Pipe {
    pub fn new(x: f32, gap_y: f32) -> Self {
        Self {
            x,
            gap_y,
            width: config::PIPE_WIDTH,
            gap: config::PIPE_GAP,
            scored: false,
        }
    }

    /// The two solid rectangles (top and bottom) used for collision.
    pub fn solids(&self, screen_h: f32) -> (Rect, Rect) {
        let top_h = self.gap_y - self.gap / 2.0;
        let bottom_y = self.gap_y + self.gap / 2.0;
        let bottom_h = screen_h - config::GROUND_HEIGHT - bottom_y;
        (
            Rect::new(self.x, 0.0, self.width, top_h),
            Rect::new(self.x, bottom_y, self.width, bottom_h),
        )
    }

    pub fn is_off_screen(&self) -> bool {
        self.x + self.width < 0.0
    }

    pub fn update(&mut self, dt: f32) {
        self.x -= config::SCROLL_SPEED * dt;
    }

    pub fn draw(&self, screen_h: f32) {
        let (top, bottom) = self.solids(screen_h);
        let green = PackedColor::new(70, 190, 70, 255).to_mq();
        let dark_green = PackedColor::new(40, 130, 40, 255).to_mq();

        for solid in &[top, bottom] {
            draw_rectangle(solid.x, solid.y, solid.w, solid.h, green);
            draw_rectangle_lines(solid.x, solid.y, solid.w, solid.h, 4.0, dark_green);
        }

        // Classic wider "lip" at each side of the gap.
        const LIP_H: f32 = 26.0;
        const OVERHANG: f32 = 6.0;
        let top_lip_y = self.gap_y - self.gap / 2.0 - LIP_H;
        let bottom_lip_y = self.gap_y + self.gap / 2.0;

        for &lip_y in &[top_lip_y, bottom_lip_y] {
            draw_rectangle(
                self.x - OVERHANG,
                lip_y,
                self.width + OVERHANG * 2.0,
                LIP_H,
                green,
            );
            draw_rectangle_lines(
                self.x - OVERHANG,
                lip_y,
                self.width + OVERHANG * 2.0,
                LIP_H,
                4.0,
                dark_green,
            );
        }
    }
}
